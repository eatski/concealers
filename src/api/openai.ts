import { OpenAI, APIError } from 'openai'

import type { Character } from '../shared'

export interface CharacterResponse {
  nextStatement: string
  urgency: number
}

function generatePromptForCharacter(currentCharacter: Character, allCharacters: Character[]) {
  const otherCharacters = allCharacters.filter(char => char !== currentCharacter)
  
  return `
あなたは「${currentCharacter.name}」というキャラクターです。

あなたの情報:
名前: ${currentCharacter.name}
説明: ${currentCharacter.description}
隠された情報: ${currentCharacter.hiddenPrompt}

他のキャラクター:
${otherCharacters.map(char => `
名前: ${char.name}
説明: ${char.description}
`).join('\n')}

この状況で、あなたが次に発言したいことを決定してください。
`.trim()
}

export async function sendRequest(apiKey: string, characters: Character[]): Promise<CharacterResponse[]> {
  if (!apiKey) return []
  
  const openai = new OpenAI({
    apiKey: apiKey,
    dangerouslyAllowBrowser: true
  })

  try {
    const responses = await Promise.all(
      characters.map(async (character) => {
        const prompt = generatePromptForCharacter(character, characters)
        
        const completion = await openai.chat.completions.create({
          model: 'gpt-4',
          messages: [
            { 
              role: 'system', 
              content: 'キャラクターの次の発言を決定し、関数を呼び出して回答してください。' 
            },
            { 
              role: 'user', 
              content: prompt 
            }
          ],
          functions: [
            {
              name: 'decideNextStatement',
              description: 'キャラクターの次の発言を決定する',
              parameters: {
                type: 'object',
                properties: {
                  nextStatement: {
                    type: 'string',
                    description: '次に発言しようとしていることを100文字程度で説明'
                  },
                  urgency: {
                    type: 'integer',
                    description: '発言への意欲（1: 消極的, 5: 積極的）',
                    minimum: 1,
                    maximum: 5
                  }
                },
                required: ['nextStatement', 'urgency']
              }
            }
          ],
          function_call: { name: 'decideNextStatement' }
        })

        const functionCall = completion.choices[0].message.function_call
        if (!functionCall || !functionCall.arguments) {
          throw new Error('APIからの応答が不正です')
        }

        try {
          const response = JSON.parse(functionCall.arguments) as CharacterResponse
          return response
        } catch (e) {
          console.error('JSON解析エラー:', e)
          throw new Error('APIからの応答の解析に失敗しました')
        }
      })
    )

    return responses
  } catch (error) {
    console.error('Error:', error)
    if (error instanceof APIError) {
      if (error.status === 401) {
        throw new Error('APIキーが無効です')
      } else if (error.status === 429) {
        throw new Error('リクエスト制限に達しました')
      }
    }
    throw error
  }
}