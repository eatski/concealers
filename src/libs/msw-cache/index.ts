import { unlink } from "node:fs/promises";
import {
  http,
  type DefaultBodyType,
  HttpResponse,
  type StrictRequest,
  bypass,
} from "msw";
import { setupServer } from "msw/node";
import { FileSystemCache } from "../file-system-cache";

interface CacheData {
  request: Record<string, unknown>;
  response: unknown;
}

class ResponseCache {
  private readonly usedCachePath = new Set<string>();
  constructor(private readonly cache: FileSystemCache) {}

  private static async configToRequestRecord(
    config: StrictRequest<DefaultBodyType>,
  ): Promise<Record<string, unknown>> {
    return {
      url: config.url,
      body: await config.text(),
      method: config.method,
    };
  }

  public async get(
    configForKey: StrictRequest<DefaultBodyType>,
  ): Promise<unknown> {
    const record = await ResponseCache.configToRequestRecord(configForKey);
    const key = JSON.stringify(record);
    this.usedCachePath.add(this.cache.getPath(key));
    const cached = await this.cache.get<CacheData>(key);
    return cached?.response;
  }

  public async set(
    configForKey: StrictRequest<DefaultBodyType>,
    value: unknown,
  ) {
    const record = await ResponseCache.configToRequestRecord(configForKey);
    const key = JSON.stringify(record);
    this.usedCachePath.add(this.cache.getPath(key));
    await this.cache.set(key, {
      request: record,
      response: value,
    });
  }

  public async clearUnusedCache() {
    const caches = await this.cache.load();
    const pathes = caches.files
      .map((file) => file.path)
      .filter((path) => !this.usedCachePath.has(path));
    await Promise.all(pathes.map(unlink));
  }
}

export const initMswCacheServer = (basePath: string) => {
  const cache = new ResponseCache(
    new FileSystemCache({
      basePath: basePath,
    }),
  );
  const server = setupServer(
    http.all("*", async ({request}) => {
      const cached = await cache.get(request.clone());
      if (typeof cached === "object") {
        return HttpResponse.json(cached);
      }
      console.log("Cache not hit. ", request.url);
    	const response = await fetch(bypass(request));
      if (response.status === 200) {
        const text = await response.clone().json();
        await cache.set(request, text);
      }
      return HttpResponse.json(await response.json());
    }),
  );
  return {
    listen: () => server.listen(),
    close: () => {
      server.close();
      return cache.clearUnusedCache();
    },
  };
};
