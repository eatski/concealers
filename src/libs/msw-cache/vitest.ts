import path from "node:path";
import filenamify from "filenamify";
import { afterAll, beforeAll } from "vitest";
import { initMswCacheServer } from ".";

export const applyTestHooks = () => {
	let server: ReturnType<typeof initMswCacheServer> | null = null;
	beforeAll(({ name, file }) => {
		if (!file?.name) {
			throw new Error("suite.file.name is not defined");
		}
		const parserdPath = path.parse(file.name);
		const filename = filenamify(name, { replacement: "_" });
		server = initMswCacheServer(
			path.resolve(
				process.cwd(),
				parserdPath.dir,
				"__data__",
				parserdPath.name + parserdPath.ext,
				filename,
			),
		);
		server.listen();
	});
	afterAll(async ({ tasks }) => {
		if (tasks.every((t) => t.result?.state === "pass")) {
			await server?.close();
		}
	});
};
