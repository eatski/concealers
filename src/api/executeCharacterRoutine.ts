import { createCharacterThoughts } from './analyzeCharacterThoughts'
import { createCharacterSpeech } from './createCharacterSpeech'
import { searchRelevantMemories } from './searchRelevantMemories'
import { OpenAI } from 'openai'
import { type Character, type RoutineResult } from '../shared'

export interface ExecuteCharacterRoutineArgs {
  openai: OpenAI
  commonPrompt: string
  characters: Character[]
  history: RoutineResult[]
  random?: () => number
}

export async function executeCharacterRoutine({
  openai,
  commonPrompt,
  characters,
  history,
  random = Math.random
}: ExecuteCharacterRoutineArgs): Promise<RoutineResult> {
  // 各キャラクターの思考を並列で生成
  const characterMemoriesPromises = characters.map(async character => {
    // キャラクターごとに関連する記憶を検索
    const relevantMemories = await searchRelevantMemories({
      openai,
      commonPrompt,
      currentCharacter: character,
      characters,
      history
    })
    
    return createCharacterThoughts({
      openai,
      commonPrompt,
      currentCharacter: character,
      allCharacters: characters,
      history,
      relevantMemories
    })
  })

  const characterMemories = await Promise.all(characterMemoriesPromises)
  const speech = await createCharacterSpeech({
    openai,
    commonPrompt,
    characters,
    characterMemories,
    history,
    random
  })
  
  return {
    characterMemories,
    speech
  }
}