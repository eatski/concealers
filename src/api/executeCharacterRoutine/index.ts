import { OpenAI } from 'openai'
import type { Character, RoutineResult } from '../../shared'
import { createCharacterThoughts } from '../analyzeCharacterThoughts'
import { createCharacterSpeech } from '../createCharacterSpeech'
import { searchRelevantMemories } from '../searchRelevantMemories'
import { selectSpeakingCharacter } from '../selectSpeakingCharacter'

export interface ExecuteCharacterRoutineArgs {
  openai: OpenAI
  commonPrompt: string
  characters: Character[]
  history: RoutineResult[]
  random?: () => number
}

export async function executeCharacterRoutine({
  openai,
  commonPrompt,
  characters,
  history,
  random = Math.random
}: ExecuteCharacterRoutineArgs): Promise<RoutineResult> {
  // 各キャラクターの思考と関連する記憶を並列で生成
  const characterData = await Promise.all(characters.map(async character => {
    // キャラクターごとに関連する記憶を検索
    const relevantMemories = await searchRelevantMemories({
      openai,
      commonPrompt,
      currentCharacter: character,
      characters,
      history
    })
    
    const memories = await createCharacterThoughts({
      openai,
      commonPrompt,
      currentCharacter: character,
      allCharacters: characters,
      history,
      relevantMemories
    })

    return {
      character,
      memories,
      relevantMemories
    }
  }))

  const characterMemories = characterData.map(data => data.memories)
  
  // 発言するキャラクターを選択
  const selectedCharacter = selectSpeakingCharacter({
    characters,
    characterMemories,
    random
  })

  const selectedData = selectedCharacter 
    ? characterData.find(data => data.character.name === selectedCharacter.name)
    : null

  const speech = selectedData
    ? await createCharacterSpeech({
        openai,
        commonPrompt,
        character: selectedData.character,
        characterMemories: selectedData.memories,
        relevantMemories: selectedData.relevantMemories,
        characters,
        history
      })
    : null
  
  return {
    characterMemories,
    speech
  }
}