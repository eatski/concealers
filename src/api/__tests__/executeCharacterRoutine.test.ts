import { describe, it, expect } from 'vitest'
import { OpenAI } from 'openai'
import seedrandom from 'seedrandom'
import { executeCharacterRoutine } from '../executeCharacterRoutine'
import type { RoutineResult } from '../../shared'
import { commonPrompt, characters, sampleHistory } from './fixtures/characterFixtures'
import { applyTestHooks } from '../../libs/msw-cache/vitest'

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
})

// テスト用の固定シード値を使用した乱数生成器
const rng = seedrandom('fixed-seed-for-test')

describe('executeCharacterRoutine', () => {
  applyTestHooks();

  it('探偵たちの思考と発言が生成される', async () => {
    const history: RoutineResult[] = []
    
    const result = await executeCharacterRoutine(
      openai,
      commonPrompt,
      characters,
      history,
      rng
    )

    // スナップショットでの検証
    expect(result).toMatchSnapshot()
  })

  it('会話履歴を考慮して思考と発言が生成される', async () => {
    const result = await executeCharacterRoutine(
      openai,
      commonPrompt,
      characters,
      sampleHistory,
      rng
    )

    // スナップショットでの検証
    expect(result).toMatchSnapshot()
  })
})