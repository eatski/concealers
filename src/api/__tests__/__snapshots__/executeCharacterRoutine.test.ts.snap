// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`executeCharacterRoutine > 会話履歴を考慮して思考と発言が生成される 1`] = `
{
  "speech": {
    "characterName": "新田探偵",
    "speech": "まあ、関係者とも話はしているが、特に目新しいことはないな。ただ、彼らも随分と動揺していた。君はどうだ？新たな手がかりは見つかったかい？",
  },
  "thoughts": [
    {
      "characterName": "新田探偵",
      "thought": "私自身も遺産が絡んでいるので情報を慎重に選びつつ、山下探偵の様子を探りたい。と同時に、他の重要な情報が出揃うのを待ちたい。",
      "urgency": 2,
    },
    {
      "characterName": "山下探偵",
      "thought": "新田探偵がどのような情報を持っているか注意深く観察し、次にどのように質問すべきか考えている。彼の反応を確認したい。",
      "urgency": 2,
    },
  ],
}
`;

exports[`executeCharacterRoutine > 探偵たちの思考と発言が生成される 1`] = `
{
  "speech": {
    "characterName": "新田探偵",
    "speech": "山下くん、人は見た目に惑わされやすいものだ。君の観察力で何か新たに気付いたことはあるかね？",
  },
  "thoughts": [
    {
      "characterName": "新田探偵",
      "thought": "山下探偵の経験不足が調査の妨げになるかもしれないが、彼の新鮮な視点には価値がある。自身の遺産に関する情報は隠しておく必要がある。",
      "urgency": 2,
    },
    {
      "characterName": "山下探偵",
      "thought": "新田探偵の行動を監視しつつ、事件の真相を探る必要がある。自身の未熟さを認識しつつも、この場で信頼を得るためには積極的に関与することが重要だと感じている。",
      "urgency": 2,
    },
  ],
}
`;
